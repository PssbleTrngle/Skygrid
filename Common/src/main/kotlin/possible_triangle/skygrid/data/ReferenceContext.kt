package possible_triangle.skygrid.data

import net.minecraft.resources.ResourceLocation
import possible_triangle.skygrid.data.xml.BlockProvider
import possible_triangle.skygrid.data.xml.Preset

open class ReferenceContext {

    private fun checkCircular(key: String) {
        //TODO
    }

    private fun contextKey(key: String): String? {
        return if (key.startsWith('#')) key.substring(1) else null
    }

    protected open fun getInternal(key: String): BlockProvider? = null

    operator fun get(key: String): BlockProvider? {
        checkCircular(key)
        return contextKey(key)?.let {
            getInternal(it)
        } ?: run {
            val id = ResourceLocation(key)
            return Preset[id]?.provider
        }
    }

    fun with(provider: BlockProvider): ReferenceContext {
        val strippedProvider = provider.stripped()
        return object : ReferenceContext() {
            override fun getInternal(key: String): BlockProvider? {
                return if (key == "provider") strippedProvider else this.getInternal(key)
            }
        }
    }

    fun with(values: Map<String, BlockProvider>): ReferenceContext {
        return object : ReferenceContext() {
            override fun getInternal(key: String): BlockProvider? {
                return values[key] ?: this.getInternal(key)
            }
        }
    }

}