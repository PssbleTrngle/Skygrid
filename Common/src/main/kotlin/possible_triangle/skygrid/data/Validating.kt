package possible_triangle.skygrid.data

import net.minecraft.core.Registry
import net.minecraft.tags.TagContainer
import net.minecraft.world.level.block.Block
import possible_triangle.skygrid.data.xml.FilterOperator

interface Validating {

    fun validate(
        blocks: Registry<Block>,
        tags: TagContainer,
        references: ReferenceContext,
        additionalFilters: List<FilterOperator> = emptyList()
    ): Boolean

}