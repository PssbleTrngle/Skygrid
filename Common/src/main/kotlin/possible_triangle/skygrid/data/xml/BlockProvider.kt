package possible_triangle.skygrid.data.xml

import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import net.minecraft.core.Registry
import net.minecraft.tags.TagContainer
import net.minecraft.world.level.block.Block
import net.minecraft.world.level.block.state.BlockState
import possible_triangle.skygrid.SkygridMod.LOGGER
import possible_triangle.skygrid.data.ReferenceContext
import possible_triangle.skygrid.data.Validating
import possible_triangle.skygrid.world.Generator
import possible_triangle.skygrid.world.IBlockAccess
import kotlin.random.Random

@Serializable
abstract class BlockProvider : WeightedEntry(), Generator<IBlockAccess>, Validating {

    protected abstract val extras: List<Extra>
    protected abstract val transformers: List<Transformer>
    protected abstract val filters: List<FilterOperator>
    abstract val name: String?

    @Transient
    lateinit var validExtras: List<Extra>

    /**
     * @return the provider without any extras
     */
    fun stripped(): BlockProvider {
        val parent = this
        val stripped = object : BlockProvider() {
            override val extras: List<Extra> = emptyList()
            override val transformers: List<Transformer> = parent.transformers
            override val filters: List<FilterOperator> = parent.filters
            override val name: String? = parent.name
            override val weight: Double
                get() = parent.weight

            override fun flat(): List<Pair<Block, Double>> = parent.flat()

            override fun internalValidate(
                blocks: Registry<Block>,
                tags: TagContainer,
                references: ReferenceContext,
                filters: List<FilterOperator>,
            ): Boolean = parent.internalValidate(blocks, tags, references, filters)

            override fun base(random: Random): Block = parent.base(random)

            override fun generateBase(random: Random, chunk: IBlockAccess): Boolean = parent.generateBase(random, chunk)
        }

        stripped.validExtras = emptyList()
        return stripped
    }

    abstract fun flat(): List<Pair<Block, Double>>

    protected abstract fun internalValidate(
        blocks: Registry<Block>,
        tags: TagContainer,
        references: ReferenceContext,
        filters: List<FilterOperator>,
    ): Boolean

    final override fun validate(
        blocks: Registry<Block>,
        tags: TagContainer,
        references: ReferenceContext,
        additionalFilters: List<FilterOperator>,
    ): Boolean {
        val referencesWithThis = references.with(this)
        validExtras = extras.filter { it.validate(blocks, tags, referencesWithThis) }
        return internalValidate(blocks, tags, references, additionalFilters + filters).also {
            if (!it) LOGGER.debug("Invalid BlockProvider ${name ?: "(anonymous)"} of type ${javaClass.name}")
        }
    }

    internal abstract fun base(random: Random): Block

    private fun getState(random: Random): BlockState {
        return transformers.fold(base(random).defaultBlockState()) { state, transformer ->
            transformer.apply(state, random)
        }
    }

    protected open fun generateBase(random: Random, chunk: IBlockAccess): Boolean {
        val state = getState(random)
        return chunk.set(state)
    }

    final override fun generate(random: Random, access: IBlockAccess): Boolean {
        val sharedSeed = random.nextLong()
        return generateBase(random, access).apply {
            if (this) validExtras.forEach {
                it.generate(if (it.shared) Random(sharedSeed) else random, access)
            }
        }
    }

}