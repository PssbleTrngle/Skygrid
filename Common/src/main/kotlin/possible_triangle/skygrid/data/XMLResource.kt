package possible_triangle.skygrid.data

import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.KSerializer
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic
import kotlinx.serialization.modules.subclass
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.MinecraftServer
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.server.packs.resources.SimplePreparableReloadListener
import net.minecraft.util.profiling.ProfilerFiller
import nl.adaptivity.xmlutil.ExperimentalXmlUtilApi
import nl.adaptivity.xmlutil.serialization.UnknownChildHandler
import nl.adaptivity.xmlutil.serialization.XML
import nl.adaptivity.xmlutil.serialization.XmlSerializationPolicy
import possible_triangle.skygrid.SkygridMod
import possible_triangle.skygrid.SkygridMod.LOGGER
import possible_triangle.skygrid.data.xml.*
import possible_triangle.skygrid.data.xml.impl.*
import java.io.File

abstract class XMLResource<T>(val path: String, private val serializer: () -> KSerializer<T>) :
    SimplePreparableReloadListener<Map<ResourceLocation, T>>() {

    @ExperimentalXmlUtilApi
    @ExperimentalSerializationApi
    companion object {
        val LOADER = XML(SerializersModule {
            polymorphic(BlockProvider::class) {
                subclass(BlockList::class)
                subclass(Fallback::class)
                subclass(Tag::class)
                subclass(Block::class)
                subclass(Reference::class)
            }

            polymorphic(Transformer::class) {
                subclass(SetPropertyTransformer::class)
                subclass(CyclePropertyTransformer::class)
            }

            polymorphic(FilterOperator::class) {
                subclass(ExceptFilter::class)
            }

            polymorphic(Extra::class) {
                subclass(Side::class)
                subclass(Offset::class)
            }

            polymorphic(Filter::class) {
                subclass(NameFilter::class)
                subclass(ModFilter::class)
                subclass(TagFilter::class)
            }
        }) {
            encodeDefault = XmlSerializationPolicy.XmlEncodeDefault.NEVER
            autoPolymorphic = true
            indentString = " ".repeat(3)
            unknownChildHandler = UnknownChildHandler { input, _, descriptor, name, candidates ->
                throw DeserializationException(
                    input.locationInfo,
                    "${descriptor.tagName}/${name ?: "<CDATA>"}",
                    candidates
                )
            }
        }

        private val RESOURCES = arrayListOf<XMLResource<*>>()

        fun reload(server: MinecraftServer) {
            RESOURCES.forEach { it.validate(server) }
        }

        fun register(register: (XMLResource<*>) -> Unit) {
            RESOURCES.forEach(register)
        }

        fun clear() {
            RESOURCES.forEach { it.values = mapOf() }
        }

    }

    open fun onReload(server: MinecraftServer) {}

    private var values = mapOf<ResourceLocation, T>()

    val keys
        get() = values.keys

    operator fun get(id: ResourceLocation): T? {
        return values[id]
    }

    fun exists(id: ResourceLocation): Boolean {
        return values.containsKey(id)
    }

    fun register() {
        RESOURCES.add(this)
    }

    private fun validate(server: MinecraftServer) {
        values = values.filterValues { validate(it, server) }
        onReload(server)
    }

    override fun prepare(manager: ResourceManager, profiler: ProfilerFiller): Map<ResourceLocation, T> {
        val resources = manager.listResources("${SkygridMod.MOD_ID}/$path") { it.endsWith(".xml") }
        LOGGER.info("Found ${resources.size} $path")
        return resources.associateWith { load(manager, it) }
            .filterValues { it != null }
            .mapValues { it.value as T }
            .mapKeys { ResourceLocation(it.key.namespace, File(it.key.path).nameWithoutExtension) }
    }

    override fun apply(loaded: Map<ResourceLocation, T>, manager: ResourceManager, profiler: ProfilerFiller) {
        LOGGER.info("Loaded ${loaded.size} skygrid $path")
        values = loaded
    }

    abstract fun merge(a: T, b: T): T

    abstract fun validate(value: T, server: MinecraftServer): Boolean

    private fun attemptDecode(id: ResourceLocation, string: String): T? {
        return try {
            LOADER.decodeFromString(serializer(), string)
        } catch (e: DeserializationException) {
            LOGGER.warn("Error loading resource '$id': Unknown field at ${e.location}: '${e.field}'")
            null
        } catch (e: Exception) {
            LOGGER.warn("Error loading resource '$id': ${e.message}")
            null
        }
    }

    private fun load(manager: ResourceManager, name: ResourceLocation): T? {
        val resources = manager.getResources(name)
        return resources.asSequence()
            .map { it.inputStream }
            .map { it.reader() }
            .map { it.readLines().joinToString(separator = "\n") }
            .mapNotNull { attemptDecode(name, it) }
            .reduceOrNull(::merge)
    }

}