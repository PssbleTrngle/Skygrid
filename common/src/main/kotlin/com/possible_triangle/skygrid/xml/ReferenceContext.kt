package com.possible_triangle.skygrid.xml

import com.possible_triangle.skygrid.api.xml.IReferenceContext
import com.possible_triangle.skygrid.api.xml.elements.BlockProvider
import net.minecraft.resources.ResourceLocation
import com.possible_triangle.skygrid.xml.resources.Presets

open class ReferenceContext : IReferenceContext {

    private fun checkCircular(key: String) {
        //TODO
    }

    private fun contextKey(key: String): String? {
        return if (key.startsWith('#')) key.substring(1) else null
    }

    protected open fun getInternal(key: String): BlockProvider? = null

    override operator fun get(key: String): BlockProvider? {
        checkCircular(key)
        return contextKey(key)?.let {
            getInternal(it)
        } ?: run {
            val id = ResourceLocation(key)
            return Presets[id]?.provider
        }
    }

    override fun with(provider: BlockProvider): ReferenceContext {
        val strippedProvider = provider.stripped()
        return object : ReferenceContext() {
            override fun getInternal(key: String): BlockProvider? {
                return if (key == "provider") strippedProvider else this.getInternal(key)
            }
        }
    }

    override fun with(values: Map<String, BlockProvider>): ReferenceContext {
        return object : ReferenceContext() {
            override fun getInternal(key: String): BlockProvider? {
                return values[key] ?: this.getInternal(key)
            }
        }
    }

}