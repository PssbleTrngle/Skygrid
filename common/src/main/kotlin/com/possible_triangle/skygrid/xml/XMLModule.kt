package com.possible_triangle.skygrid.xml

import com.possible_triangle.skygrid.api.events.RegisterElementEvent
import com.possible_triangle.skygrid.api.events.register
import com.possible_triangle.skygrid.api.events.subclass
import com.possible_triangle.skygrid.api.xml.DeserializationException
import com.possible_triangle.skygrid.api.xml.elements.*
import com.possible_triangle.skygrid.api.xml.elements.extras.Cardinal
import com.possible_triangle.skygrid.api.xml.elements.extras.Offset
import com.possible_triangle.skygrid.api.xml.elements.extras.Side
import com.possible_triangle.skygrid.api.xml.elements.extras.Surround
import com.possible_triangle.skygrid.api.xml.elements.filters.ExceptFilter
import com.possible_triangle.skygrid.api.xml.elements.filters.ModFilter
import com.possible_triangle.skygrid.api.xml.elements.filters.NameFilter
import com.possible_triangle.skygrid.api.xml.elements.filters.TagFilter
import com.possible_triangle.skygrid.api.xml.elements.providers.*
import com.possible_triangle.skygrid.api.xml.elements.transformers.CyclePropertyTransformer
import com.possible_triangle.skygrid.api.xml.elements.transformers.SetPropertyTransformer
import kotlinx.serialization.KSerializer
import kotlinx.serialization.modules.PolymorphicModuleBuilder
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.SerializersModuleBuilder
import kotlinx.serialization.modules.polymorphic
import nl.adaptivity.xmlutil.ExperimentalXmlUtilApi
import nl.adaptivity.xmlutil.serialization.UnknownChildHandler
import nl.adaptivity.xmlutil.serialization.XML
import nl.adaptivity.xmlutil.serialization.XmlSerializationPolicy
import kotlin.reflect.KClass

private typealias Sub<T> = Pair<KClass<T>, KSerializer<T>>

private fun SerializersModuleBuilder.collectRegisteredElements() {
    val collected = hashMapOf<KClass<*>, MutableCollection<Sub<*>>>()

    RegisterElementEvent.EVENT.invoke(object : RegisterElementEvent {
        override fun <TParent : Any, TSub : TParent> register(
            parent: KClass<TParent>,
            subclass: KClass<TSub>,
            serializer: KSerializer<TSub>,
        ) {
            collected.getOrPut(parent, ::hashSetOf).add(subclass to serializer)
        }
    })

    fun <TParent : Any, TSub : TParent> PolymorphicModuleBuilder<TParent>.subclass(it: Sub<*>) {
        subclass(it.first as KClass<TSub>, it.second as KSerializer<TSub>)
    }

    fun <TParent : Any> registerPolymorphic(parent: KClass<TParent>, subclasses: Collection<Sub<*>>) {
        polymorphic(parent) {
            subclasses.forEach {
                subclass(it)
            }
        }
    }

    collected.forEach { (parent, subclasses) ->
        registerPolymorphic(parent, subclasses)
    }
}

fun RegisterElementEvent.registerDefaultElements() {
    register(BlockProvider::class) {
        subclass(BlockList::class)
        subclass(Fallback::class)
        subclass(Tag::class)
        subclass(SingleBlock::class)
        subclass(Reference::class)
    }

    register(Transformer::class) {
        subclass(SetPropertyTransformer::class)
        subclass(CyclePropertyTransformer::class)
    }

    register(FilterOperator::class) {
        subclass(ExceptFilter::class)
    }

    register(Extra::class) {
        subclass(Side::class)
        subclass(Cardinal::class)
        subclass(Offset::class)
        subclass(Surround::class)
    }

    register(Filter::class) {
        subclass(NameFilter::class)
        subclass(ModFilter::class)
        subclass(TagFilter::class)
    }
}

@ExperimentalXmlUtilApi
fun createXMLModule() = XML(SerializersModule {
    collectRegisteredElements()
}) {
    defaultPolicy {
        indentString = " ".repeat(3)
        autoPolymorphic = true
        encodeDefault = XmlSerializationPolicy.XmlEncodeDefault.NEVER
        unknownChildHandler = UnknownChildHandler { input, _, descriptor, name, candidates ->
            throw DeserializationException(
                input.locationInfo,
                "${descriptor.tagName}/${name ?: "<CDATA>"}",
                candidates
            )
        }
    }
}