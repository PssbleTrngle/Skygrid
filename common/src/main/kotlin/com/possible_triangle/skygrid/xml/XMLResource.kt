package com.possible_triangle.skygrid.xml

import com.possible_triangle.skygrid.SkygridMod.LOGGER
import com.possible_triangle.skygrid.api.SkygridConstants
import com.possible_triangle.skygrid.api.xml.DeserializationException
import com.possible_triangle.skygrid.api.xml.elements.*
import com.possible_triangle.skygrid.api.xml.elements.impl.*
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.KSerializer
import kotlinx.serialization.modules.SerializersModule
import kotlinx.serialization.modules.polymorphic
import kotlinx.serialization.modules.subclass
import net.minecraft.resources.ResourceLocation
import net.minecraft.server.MinecraftServer
import net.minecraft.server.packs.resources.Resource
import net.minecraft.server.packs.resources.ResourceManager
import net.minecraft.server.packs.resources.SimplePreparableReloadListener
import net.minecraft.util.profiling.ProfilerFiller
import nl.adaptivity.xmlutil.ExperimentalXmlUtilApi
import nl.adaptivity.xmlutil.serialization.UnknownChildHandler
import nl.adaptivity.xmlutil.serialization.XML
import nl.adaptivity.xmlutil.serialization.XmlSerializationPolicy
import java.io.File

abstract class XMLResource<T>(val path: String, private val serializer: () -> KSerializer<T>) :
    SimplePreparableReloadListener<Map<ResourceLocation, T>>() {

    @ExperimentalXmlUtilApi
    @ExperimentalSerializationApi
    companion object {
        val LOADER = XML(SerializersModule {
            polymorphic(BlockProvider::class) {
                subclass(BlockList::class)
                subclass(Fallback::class)
                subclass(Tag::class)
                subclass(SingleBlock::class)
                subclass(Reference::class)
            }

            polymorphic(Transformer::class) {
                subclass(SetPropertyTransformer::class)
                subclass(CyclePropertyTransformer::class)
            }

            polymorphic(FilterOperator::class) {
                subclass(ExceptFilter::class)
            }

            polymorphic(Extra::class) {
                subclass(Side::class)
                subclass(Cardinal::class)
                subclass(Offset::class)
            }

            polymorphic(Filter::class) {
                subclass(NameFilter::class)
                subclass(ModFilter::class)
                subclass(TagFilter::class)
            }
        }) {
            defaultPolicy {
                indentString = " ".repeat(3)
                autoPolymorphic = true
                encodeDefault = XmlSerializationPolicy.XmlEncodeDefault.NEVER
                unknownChildHandler = UnknownChildHandler { input, _, descriptor, name, candidates ->
                    throw DeserializationException(
                        input.locationInfo,
                        "${descriptor.tagName}/${name ?: "<CDATA>"}",
                        candidates
                    )
                }
            }
        }

        private val RESOURCES = arrayListOf<XMLResource<*>>()

        fun reload(server: MinecraftServer) {
            RESOURCES.forEach { it.validate(server) }
        }

        fun register(register: (XMLResource<*>) -> Unit) {
            RESOURCES.forEach(register)
        }

        fun clear() {
            RESOURCES.forEach { it.values = mapOf() }
        }

    }

    open fun onReload(server: MinecraftServer) {}

    private var values = mapOf<ResourceLocation, T>()

    val keys
        get() = values.keys

    val entries
        get() = values.entries

    operator fun get(id: ResourceLocation): T? {
        return values[id]
    }

    fun exists(id: ResourceLocation): Boolean {
        return values.containsKey(id)
    }

    fun register() {
        RESOURCES.add(this)
    }

    private fun validate(server: MinecraftServer) {
        values = values.filterValues { validate(it, server) }
        onReload(server)
    }

    override fun prepare(manager: ResourceManager, profiler: ProfilerFiller): Map<ResourceLocation, T> {
        val resources = manager.listResourceStacks("${SkygridConstants.MOD_ID}/$path") { it.path.endsWith(".xml") }
        LOGGER.info("Found ${resources.size} skygrid $path")
        return resources.mapValues { load(it.key, it.value) }
            .filterValues { it != null }
            .mapValues { it.value as T }
            .mapKeys { ResourceLocation(it.key.namespace, File(it.key.path).nameWithoutExtension) }
    }

    override fun apply(loaded: Map<ResourceLocation, T>, manager: ResourceManager, profiler: ProfilerFiller) {
        LOGGER.info("Loaded ${loaded.size} skygrid $path")
        values = loaded
    }

    abstract fun merge(a: T, b: T): T

    abstract fun validate(value: T, server: MinecraftServer): Boolean

    private fun attemptDecode(id: ResourceLocation, string: String): T? {
        return try {
            LOADER.decodeFromString(serializer(), string)
        } catch (e: DeserializationException) {
            LOGGER.warn("Error loading resource '$id': Unknown field at ${e.location}: '${e.field}'")
            null
        } catch (e: Exception) {
            LOGGER.warn("Error loading resource '$id': ${e.message}")
            null
        }
    }

    private fun load(name: ResourceLocation, resources: List<Resource>): T? {
        return resources.asSequence()
            .map { it.open() }
            .map { it.reader() }
            .map { it.readLines().joinToString(separator = "\n") }
            .mapNotNull { attemptDecode(name, it) }
            .reduceOrNull(::merge)
    }

}