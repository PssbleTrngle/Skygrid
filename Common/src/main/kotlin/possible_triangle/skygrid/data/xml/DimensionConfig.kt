package possible_triangle.skygrid.data.xml

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.Transient
import net.minecraft.core.Registry
import net.minecraft.core.RegistryAccess
import net.minecraft.nbt.CompoundTag
import net.minecraft.nbt.NbtOps
import net.minecraft.server.MinecraftServer
import net.minecraft.tags.TagContainer
import net.minecraft.util.random.SimpleWeightedRandomList
import net.minecraft.world.level.SpawnData
import net.minecraft.world.level.block.Blocks
import net.minecraft.world.level.block.EntityBlock
import net.minecraft.world.level.block.entity.BlockEntityType
import nl.adaptivity.xmlutil.serialization.XmlSerialName
import possible_triangle.skygrid.SkygridMod
import possible_triangle.skygrid.data.XMLResource
import possible_triangle.skygrid.data.xml.impl.LootTable
import possible_triangle.skygrid.data.xml.impl.SpawnerEntry
import possible_triangle.skygrid.world.BlockAccess
import possible_triangle.skygrid.world.Generator
import java.util.*
import kotlin.random.Random
import possible_triangle.skygrid.data.xml.impl.Block as SingleBlock

@Serializable
@SerialName("dimension")
data class DimensionConfig(
    @XmlSerialName("blocks", "", "") val blocks: ListWrapper<BlockProvider>,
    @XmlSerialName("loot", "", "") val loot: ListWrapper<LootTable> = ListWrapper(),
    @XmlSerialName("mobs", "", "") val mobs: ListWrapper<SpawnerEntry> = ListWrapper(),
    val replace: Boolean = false,
    val endPortals: Boolean = false,
    val minY: Int = Int.MIN_VALUE,
    val maxY: Int = 100,
    val distance: Distance = Distance(4, 4, 4),
    @XmlSerialName("gap", "", "") private val unsafeGap: SingleBlock? = null,
) : Generator<BlockAccess> {

    @Transient
    lateinit var gap: Optional<SingleBlock>

    fun validate(registries: RegistryAccess, tags: TagContainer): Boolean {
        val blockRegistry = registries.registryOrThrow(Registry.BLOCK_REGISTRY)
        val entityRegistry = registries.registryOrThrow(Registry.ENTITY_TYPE_REGISTRY)
        loot.validate { true }
        mobs.validate { entityRegistry.containsKey(it.key) }
        gap = Optional.ofNullable(unsafeGap).filter { it.validate(blockRegistry, tags) }
        return blocks.validate { it.validate(blockRegistry, tags) }
    }

    override fun generate(random: Random, access: BlockAccess): Boolean {
        val generateLoot = loot.isValid()
        val fillSpawners = mobs.isValid()

        return this.blocks.random(random).generate(random) { state, pos ->
            if (access.set(state, pos)) {
                val block = state.block

                if (block is EntityBlock) {
                    val nbt = if (generateLoot && state.`is`(SkygridMod.LOOT_CONTAINERS)) {
                        CompoundTag().apply {
                            putString("id", "mob_spawner")
                            val lootTable = loot.random(random)
                            putString("LootTable", lootTable.id)
                            putLong("LootTableSeed", random.nextLong())
                        }
                    } else if (fillSpawners && state.`is`(Blocks.SPAWNER)) {
                        CompoundTag().apply {
                            val mob = mobs.random(random)
                            val data = mob.createSpawnData()
                            val potentials = SimpleWeightedRandomList.builder<SpawnData>().add(data, 1).build()

                            SpawnData.CODEC.encodeStart(NbtOps.INSTANCE, data).result().ifPresent {
                                put("SpawnData", it)
                            }

                            SpawnData.LIST_CODEC.encodeStart(NbtOps.INSTANCE, potentials).result().ifPresent {
                                put("SpawnPotentials", it)
                            }
                        }
                    } else null

                    if (nbt != null) block.newBlockEntity(pos, state)?.also {
                        nbt.putString("id", BlockEntityType.getKey(it.type).toString())
                        access.setNBT(pos, nbt)
                    }
                }

                true
            } else {
                false
            }
        }
    }

    companion object : XMLResource<DimensionConfig>("dimensions", { DimensionConfig.serializer() }) {

        val DEFAULT = DimensionConfig(ListWrapper(SingleBlock("bedrock")))

        override fun merge(a: DimensionConfig, b: DimensionConfig): DimensionConfig {
            return if (b.replace) b
            else b.copy(
                replace = false,
                blocks = a.blocks + b.blocks,
                loot = a.loot + b.loot,
                mobs = a.mobs + b.mobs,
            )
        }

        override fun onReload(server: MinecraftServer) {
            validate(DEFAULT, server)
        }

        override fun validate(value: DimensionConfig, server: MinecraftServer): Boolean {
            return value.validate(server.registryAccess(), server.tags)
        }

    }

}