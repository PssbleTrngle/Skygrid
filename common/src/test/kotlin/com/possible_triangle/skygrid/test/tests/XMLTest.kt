package com.possible_triangle.skygrid.test.tests

import au.com.origin.snapshots.Expect
import au.com.origin.snapshots.junit5.SnapshotExtension
import com.possible_triangle.skygrid.test.TestExtension
import kotlinx.serialization.ExperimentalSerializationApi
import kotlinx.serialization.encodeToString
import nl.adaptivity.xmlutil.ExperimentalXmlUtilApi
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import possible_triangle.skygrid.data.XMLResource
import possible_triangle.skygrid.data.xml.DimensionConfig
import possible_triangle.skygrid.data.xml.ListWrapper
import possible_triangle.skygrid.data.xml.impl.*

@ExperimentalSerializationApi
@ExperimentalXmlUtilApi
@ExtendWith(TestExtension::class, SnapshotExtension::class)
class XMLTest {

    private lateinit var expect: Expect

    @Test
    fun encodesCorrectly() {
        val config = DimensionConfig(
            blocks = ListWrapper(
                SingleBlock(
                    "some_block", "some_mod", extras = listOf(
                        Side(
                            "down", listOf(
                                SingleBlock("something_below"),
                            )
                        )
                    )
                ),
                SingleBlock("another_block", weight = 0.1),
                Fallback(
                    children = listOf(
                        Tag("type/material", "forge"),
                        Reference("a_preset", weight = 12.5),
                    ),
                    filters = listOf(
                        ExceptFilter(
                            listOf(
                                TagFilter("exclude_this", "a_mod"),
                                TagFilter("exclude_that"),
                                NameFilter("_stripped"),
                                ModFilter("another_mod"),
                            )
                        )
                    )
                ),
                BlockList(
                    children = listOf(
                        Tag("stuff"),
                        BlockList(
                            children = listOf(
                                SingleBlock("yet_another_block", weight = 1.0),
                                SingleBlock(
                                    "maybe_a_crafting_table", weight = 24.2,
                                    transformers = listOf(
                                        SetPropertyTransformer("numerical_property", "23"),
                                    )
                                ),
                            )
                        )
                    ),
                    transformers = listOf(
                        SetPropertyTransformer("the_key", "some_value"),
                        CyclePropertyTransformer("axis_or_something"),
                    )
                )
            )
        )
        val encoded = XMLResource.LOADER.encodeToString(config)
        expect.toMatchSnapshot(encoded)
    }

}