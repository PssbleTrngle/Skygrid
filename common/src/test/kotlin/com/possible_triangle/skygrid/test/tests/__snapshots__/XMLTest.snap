com.possible_triangle.skygrid.test.tests.XMLTest.encodesCorrectly=[
<dimension>
   <blocks>
      <children type="block">
         <value id="some_block" mod="some_mod">
            <extras type="side">
               <value on="down">
                  <providers type="block">
                     <value id="something_below"/>
                  </providers>
               </value>
            </extras>
         </value>
      </children>
      <children type="block">
         <value id="another_block" weight="0.1"/>
      </children>
      <children type="fallback">
         <value>
            <filters type="except">
               <value>
                  <filters type="tag">
                     <value id="exclude_this" mod="a_mod"/>
                  </filters>
                  <filters type="tag">
                     <value id="exclude_that"/>
                  </filters>
                  <filters type="name">
                     <value pattern="_stripped"/>
                  </filters>
                  <filters type="mod">
                     <value id="another_mod"/>
                  </filters>
               </value>
            </filters>
            <children type="tag">
               <value id="type/material" mod="forge"/>
            </children>
            <children type="reference">
               <value id="a_preset" weight="12.5"/>
            </children>
         </value>
      </children>
      <children type="list">
         <value>
            <transformers type="set">
               <value key="the_key" value="some_value"/>
            </transformers>
            <transformers type="cycle">
               <value key="axis_or_something"/>
            </transformers>
            <children type="tag">
               <value id="stuff"/>
            </children>
            <children type="list">
               <value>
                  <children type="block">
                     <value id="yet_another_block"/>
                  </children>
                  <children type="block">
                     <value id="maybe_a_crafting_table" weight="24.2">
                        <transformers type="set">
                           <value key="numerical_property" value="23"/>
                        </transformers>
                     </value>
                  </children>
               </value>
            </children>
         </value>
      </children>
   </blocks>
</dimension>
]